File: mobile/android/app/src/main/java/ai/onedeck/mobile/OneDeckJsBridge.kt
package ai.onedeck.mobile

import android.Manifest
import android.content.pm.PackageManager
import android.util.Log
import android.webkit.JavascriptInterface
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import org.json.JSONObject

/**
 * JSAPI Bridge
 * 原生能力暴露给前端 WebView，前端通过 window.OneDeckNative 调用
 *
 * 安全限制：只暴露读取能力，不暴露写入/修改系统状态的能力
 */
class OneDeckJsBridge(private val activity: MainActivity) {

    private val tag = "OneDeckJsBridge"
    private val pendingPermissionCallbacks = mutableMapOf<Int, (Boolean) -> Unit>()

    /**
     * 调用 JSAPI
     * 由前端 JS 调用：window.OneDeckNative.callApi(apiName, argsJson)
     */
    @JavascriptInterface
    fun callApi(apiName: String, argsJson: String): String {
        val args = try {
            JSONObject(argsJson)
        } catch (e: Exception) {
            return errorResult("Invalid JSON args")
        }

        val result = when (apiName) {
            // 文件操作
            "file.read" -> handleFileRead(args)
            "file.write" -> handleFileWrite(args)
            "file.delete" -> handleFileDelete(args)
            "file.exists" -> handleFileExists(args)
            "file.list" -> handleFileList(args)

            // 设备信息
            "device.getInfo" -> handleGetDeviceInfo()
            "device.getBattery" -> handleGetBattery()
            "device.vibrate" -> handleVibrate(args)

            // 剪贴板
            "clipboard.read" -> handleClipboardRead()
            "clipboard.write" -> handleClipboardWrite(args)

            // 网络
            "network.getStatus" -> handleGetNetworkStatus()

            // 摄像头
            "camera.takePhoto" -> handleTakePhoto()
            "camera.startStream" -> handleStartStream()
            "camera.stopStream" -> handleStopStream(args)

            // 定位
            "location.getCurrent" -> handleGetCurrentLocation()
            "location.startWatch" -> handleStartLocationWatch()
            "location.stopWatch" -> handleStopLocationWatch(args)

            // 存储
            "storage.get" -> handleStorageGet(args)
            "storage.set" -> handleStorageSet(args)
            "storage.remove" -> handleStorageRemove(args)
            "storage.keys" -> handleStorageKeys(args)
            "storage.clear" -> handleStorageClear(args)

            else -> errorResult("Unknown API: $apiName")
        }

        return result
    }

    /**
     * 获取设备 ID
     */
    @JavascriptInterface
    fun getDeviceId(): String {
        return android.provider.Settings.Secure.getString(
            activity.contentResolver,
            android.provider.Settings.Secure.ANDROID_ID
        )
    }

    /**
     * 获取平台
     */
    @JavascriptInterface
    fun getPlatform(): String = "android"

    /**
     * 请求权限
     */
    @JavascriptInterface
    fun requestPermission(permission: String): Boolean {
        val androidPermission = when (permission) {
            "camera" -> Manifest.permission.CAMERA
            "location" -> Manifest.permission.ACCESS_FINE_LOCATION
            "storage" -> Manifest.permission.WRITE_EXTERNAL_STORAGE
            else -> return false
        }

        val granted = ContextCompat.checkSelfPermission(activity, androidPermission) ==
                PackageManager.PERMISSION_GRANTED

        if (!granted) {
            ActivityCompat.requestPermissions(activity, arrayOf(androidPermission), 1001)
        }

        return granted
    }

    /**
     * 记录日志（转发到桌面端）
     */
    @JavascriptInterface
    fun log(level: String, message: String) {
        when (level) {
            "debug" -> Log.d(tag, message)
            "info" -> Log.i(tag, message)
            "warn" -> Log.w(tag, message)
            "error" -> Log.e(tag, message)
        }
    }

    // ==================== JSAPI 实现占位 ====================

    private fun handleFileRead(args: JSONObject): String {
        // TODO: 在插件沙箱目录内读取文件
        return successResult(mapOf("content" to ""))
    }

    private fun handleFileWrite(args: JSONObject): String {
        // TODO: 在插件沙箱目录内写入文件
        return successResult(null)
    }

    private fun handleFileDelete(args: JSONObject): String {
        return successResult(null)
    }

    private fun handleFileExists(args: JSONObject): String {
        return successResult(mapOf("exists" to false))
    }

    private fun handleFileList(args: JSONObject): String {
        return successResult(mapOf("files" to emptyList<String>()))
    }

    private fun handleGetDeviceInfo(): String {
        return successResult(mapOf(
            "platform" to "android",
            "osVersion" to android.os.Build.VERSION.RELEASE,
            "model" to android.os.Build.MODEL,
            "screenResolution" to mapOf(
                "width" to activity.resources.displayMetrics.widthPixels,
                "height" to activity.resources.displayMetrics.heightPixels
            )
        ))
    }

    private fun handleGetBattery(): String {
        // TODO: 使用 BatteryManager 获取电量
        return successResult(mapOf("level" to 100, "charging" to false))
    }

    private fun handleVibrate(args: JSONObject): String {
        // TODO: 使用 Vibrator 触发震动
        return successResult(null)
    }

    private fun handleClipboardRead(): String {
        // TODO: 读取剪贴板
        return successResult(mapOf("text" to ""))
    }

    private fun handleClipboardWrite(args: JSONObject): String {
        // TODO: 写入剪贴板
        return successResult(null)
    }

    private fun handleGetNetworkStatus(): String {
        // TODO: 使用 ConnectivityManager 获取网络状态
        return successResult(mapOf("connected" to true, "type" to "wifi"))
    }

    private fun handleTakePhoto(): String {
        // TODO: 调用摄像头拍照
        return successResult(mapOf("data" to ""))
    }

    private fun handleStartStream(): String {
        return successResult(mapOf("streamId" to ""))
    }

    private fun handleStopStream(args: JSONObject): String {
        return successResult(null)
    }

    private fun handleGetCurrentLocation(): String {
        // TODO: 使用 FusedLocationProviderClient 获取位置
        return successResult(mapOf("lat" to 0.0, "lng" to 0.0, "accuracy" to 0.0))
    }

    private fun handleStartLocationWatch(): String {
        return successResult(mapOf("watchId" to ""))
    }

    private fun handleStopLocationWatch(args: JSONObject): String {
        return successResult(null)
    }

    private fun handleStorageGet(args: JSONObject): String {
        // TODO: 从 SQLite 读取插件数据
        return successResult(mapOf("value" to null))
    }

    private fun handleStorageSet(args: JSONObject): String {
        // TODO: 写入 SQLite 插件数据
        return successResult(null)
    }

    private fun handleStorageRemove(args: JSONObject): String {
        return successResult(null)
    }

    private fun handleStorageKeys(args: JSONObject): String {
        return successResult(mapOf("keys" to emptyList<String>()))
    }

    private fun handleStorageClear(args: JSONObject): String {
        return successResult(null)
    }

    // ==================== 辅助方法 ====================

    private fun successResult(data: Any?): String {
        val result = JSONObject()
        result.put("success", true)
        if (data != null) result.put("data", data)
        return result.toString()
    }

    private fun errorResult(message: String): String {
        val result = JSONObject()
        result.put("success", false)
        result.put("error", message)
        return result.toString()
    }
}

File: mobile/ios/OneDeckMobile/OneDeckJsBridge.swift
import Foundation
import WebKit

/**
 * JSAPI Bridge
 * 暴露设备能力给前端 WebView
 * 前端通过 window.OneDeckNative.callApi(apiName, args) 调用
 *
 * 安全限制：只暴露读取能力，不暴露写入/修改系统状态的能力
 */
class OneDeckJsBridge: NSObject, WKScriptMessageHandler {
    private let logBuffer = LogBuffer(maxSize: 1000)

    func userContentController(_ userContentController: WKUserContentController, didReceive message: WKScriptMessage) {
        guard message.name == "OneDeckNative" else { return }

        guard let body = message.body as? [String: Any],
              let method = body["method"] as? String else {
            return
        }

        switch method {
        case "callApi":
            handleCallApi(body)
        case "getDeviceId":
            handleGetDeviceId(message.webView)
        case "getPlatform":
            handleGetPlatform(message.webView)
        case "requestPermission":
            handleRequestPermission(body)
        case "log":
            handleLog(body)
        default:
            break
        }
    }

    // MARK: - JSAPI 实现

    private func handleCallApi(_ body: [String: Any]) {
        guard let apiName = body["apiName"] as? String,
              let argsJson = body["args"] as? String else { return }

        let args = parseJSON(argsJson) ?? [:]
        var result: [String: Any] = ["success": true]

        switch apiName {
        case "device.getInfo":
            result["data"] = [
                "platform": "ios",
                "osVersion": UIDevice.current.systemVersion,
                "model": UIDevice.current.model,
                "screenResolution": [
                    "width": UIScreen.main.bounds.width,
                    "height": UIScreen.main.bounds.height
                ]
            ]

        case "device.getBattery":
            // TODO: UIDevice.current.batteryLevel
            result["data"] = ["level": 100, "charging": false]

        case "device.vibrate":
            // TODO: UIImpactFeedbackGenerator
            break

        case "clipboard.read":
            // TODO: UIPasteboard
            result["data"] = ["text": ""]

        case "clipboard.write":
            // TODO: UIPasteboard
            break

        case "network.getStatus":
            // TODO: NWPathMonitor
            result["data"] = ["connected": true, "type": "wifi"]

        case "file.read", "file.write", "file.delete", "file.exists", "file.list":
            // TODO: 在插件沙箱目录操作
            break

        case "camera.takePhoto", "camera.startStream", "camera.stopStream":
            // TODO: AVFoundation
            break

        case "location.getCurrent", "location.startWatch", "location.stopWatch":
            // TODO: CoreLocation
            break

        case "storage.get", "storage.set", "storage.remove", "storage.keys", "storage.clear":
            // TODO: SQLite 插件数据
            break

        default:
            result = ["success": false, "error": "Unknown API: \(apiName)"]
        }
    }

    private func handleGetDeviceId(_ webView: WKWebView?) {
        let deviceId = UIDevice.current.identifierForVendor?.uuidString ?? ""
        webView?.evaluateJavaScript("window.__onedeck_deviceId = '\(deviceId)'")
    }

    private func handleGetPlatform(_ webView: WKWebView?) {
        webView?.evaluateJavaScript("window.__onedeck_platform = 'ios'")
    }

    private func handleRequestPermission(_ body: [String: Any]) {
        // TODO: 请求权限并记录日志
    }

    private func handleLog(_ body: [String: Any]) {
        let level = body["level"] as? String ?? "info"
        let message = body["message"] as? String ?? ""

        let logEntry: [String: Any] = [
            "level": level,
            "message": message,
            "timestamp": Int(Date().timeIntervalSince1970 * 1000)
        ]

        if let jsonData = try? JSONSerialization.data(withJSONObject: logEntry),
           let jsonStr = String(data: jsonData, encoding: .utf8) {
            logBuffer.add(jsonStr)
        }

        // 如果已连接，立即发送到桌面端
        // TODO: 通过 WebSocket 发送
    }

    // MARK: - 辅助

    private func parseJSON(_ str: String) -> [String: Any]? {
        guard let data = str.data(using: .utf8) else { return nil }
        return try? JSONSerialization.jsonObject(with: data) as? [String: Any]
    }
}

/// 日志环形缓冲区
class LogBuffer {
    private let maxSize: Int
    private var buffer: [String] = []

    init(maxSize: Int = 1000) {
        self.maxSize = maxSize
    }

    func add(_ entry: String) {
        if buffer.count >= maxSize {
            buffer.removeFirst()
        }
        buffer.append(entry)
    }

    func drain() -> [String] {
        let logs = buffer
        buffer.removeAll()
        return logs
    }
}
